gpflow/conditionals.py

```diff
@@ -122,6 +232,267 @@ def base_conditional(Kmn, Kmm, Knn, f, *, full_cov=False, q_sqrt=None, white=Fal
+    # TODO: Allow broadcasting over L if priors are shared?
+    # TODO: Change Kmn to be L x M x N x P? Saves a transpose...
+    R = tf.shape(f)[2]
+    M, L, N, P = [tf.shape(Kmn)[i] for i in range(Kmn.shape.ndims)]
+
+    Lm = tf.cholesky(Kmm)  # L x M x M
+
+    # Compute the projection matrix A
+    Kmn = tf.reshape(tf.transpose(Kmn, (1, 0, 2, 3)), (L, M, N * P))
+    A = tf.matrix_triangular_solve(Lm, Kmn, lower=True)  # L x M x M  *  L x M x NP  ->  L x M x NP
+    Ar = tf.reshape(A, (L, M, N, P))
+
+    # compute the covariance due to the conditioning
+    if full_cov and full_cov_output:
+        fvar = Knn - tf.tensordot(Ar, Ar, [[0, 1], [0, 1]])  # N x P x N x P
+    elif full_cov and not full_cov_output:
+        At = tf.reshape(tf.transpose(Ar), (P, N, M * L))  # P x N x ML
+        fvar = Knn - tf.matmul(At, At, transpose_b=True)  # P x N x N
+    elif not full_cov and full_cov_output:
+        At = tf.reshape(tf.transpose(Ar, [2, 3, 1, 0]), (N, P, M * L))  # N x P x ML
+        fvar = Knn - tf.matmul(At, At, transpose_b=True)  # N x P x P
+    elif not full_cov and not full_cov_output:
+        fvar = Knn - tf.reshape(tf.reduce_sum(tf.square(A), [0, 1]), (N, P))  # Knn: N x P
+
+    # another backsubstitution in the unwhitened case
+    if not white:
+        A = tf.matrix_triangular_solve(Lm, Ar)  # L x M x M  *  L x M x NP  ->  L x M x NP
+        Ar = tf.reshape(A, (L, M, N, P))
+
+    # mean: N x R x P
+    fmean = tf.tensordot(f, Ar, [[0, 1], [1, 0]])  # R x NP
+    fmean = tf.transpose(tf.reshape(fmean, (R, N, P)), [1, 0, 2])  # N x R x P
+
+    if q_sqrt is not None:
+        Lf = tf.matrix_band_part(q_sqrt, -1, 0)  # R x L x M x M
+        if q_sqrt.shape.ndims == 4:
+            # Broadcast over L?
+            A_tiled = tf.tile(A[None, :, :, :], tf.stack([R, 1, 1, 1]))  # R x L x M x NP
+            LTA = tf.matmul(Lf, A_tiled, transpose_a=True)  # R x L x M x M  *  R x L x M x NP  ->  R x L x M x NP
+        else:
+            raise NotImplementedError()
+
+        if full_cov and full_cov_output:
+            LTAr = tf.reshape(LTA, (R, L * M, N * P))
+            fvar = fvar[None, :, :, :, :] + tf.reshape(tf.matmul(LTAr, LTAr, transpose_a=True), (R, N, P, N, P))
+        elif full_cov and not full_cov_output:
+            LTAr = tf.transpose(tf.reshape(LTA, (R, L * M, N, P)), [0, 3, 1, 2])  # R x P x LM x N
+            fvar = fvar[None, :, :, :] + tf.matmul(LTAr, LTAr, transpose_a=True)  # R x P x N x N
+        elif not full_cov and full_cov_output:
+            LTAr = tf.transpose(tf.reshape(LTA, (R, L * M, N, P)), [2, 0, 1, 3])  # N x R x LM x P
+            fvar = fvar[:, None, :, :] + tf.matmul(LTAr, LTAr, transpose_a=True)  # N x R x P x P
+        elif not full_cov and not full_cov_output:
+            # N x R x P
+            fvar = fvar[None, :, :] + tf.reshape(tf.reduce_sum(tf.square(LTA), (1, 2)), (R, N, P))
+            fvar = tf.transpose(fvar, (1, 0, 2))
+    return fmean[:, 0, :], fvar[:, 0, :]
```

Comment:
why only the first element? is this because it's length 1 along that dimension and that's supposed to be a squeeze?